code/tests/unittests/test_vector3.cpp

```diff
@@ -0,0 +1,158 @@
+#include "vector3.h"
+
+#include "gmock/gmock.h"
+
+using namespace testing;
+
+class Vector3Test : public ::testing::Test {
+protected:
+    Vector3 test_vector;
+
+    void SetUp() override {
+        test_vector = Vector3(1.0, 1.0, 1.0);
+    }
+};
+
+TEST_F(Vector3Test, OutOperatorWhenPassingAStreamAndVector3OutputsTheVector3Content)
+{
+    std::string expected_value = "1 1 1";
+    std::stringstream actual_stream;
+
+    actual_stream << test_vector;
+
+    ASSERT_EQ(actual_stream.str(), expected_value);
+}
+
+TEST_F(Vector3Test, PlusOperatorWhenAddingUpTwoVector3ReturnsTheSumOfBothValues) {
+    Vector3 expected_vector(2.0, 2.0, 2.0);
+
+    Vector3 actualVector = test_vector + test_vector;
+
+    ASSERT_THAT(actualVector.e, ElementsAreArray(expected_vector.e));
+}
+
+TEST_F(Vector3Test, PlusEqualOperatorWhenAddingUpAVector3ReturnsTheSumOfBothValues) {
+    Vector3 expected_vector(2.0, 2.0, 2.0);
+    Vector3 actual_vector = test_vector;
+
+    actual_vector += test_vector;
+
+    ASSERT_THAT(actual_vector.e, ElementsAreArray(expected_vector.e));
+}
+
+TEST_F(Vector3Test, MinusOperatorWhenSubtractingTwoVector3ReturnsTheSubtractionOfBothValues) {
+    Vector3 expected_vector(0.0, 0.0, 0.0);
+
+    Vector3 actualVector = test_vector-test_vector;
+
+    ASSERT_THAT(actualVector.e, ElementsAreArray(expected_vector.e));
+}
+
+TEST_F(Vector3Test, MinusOperatorWhenNoParamsReturnsInvertedVector3) {
+    Vector3 expected_vector(-1.0, -1.0, -1.0);
+
+    Vector3 actualVector = -test_vector;
+
+    ASSERT_THAT(actualVector.e, ElementsAreArray(expected_vector.e));
+}
+
+TEST_F(Vector3Test, BracketOperatorWhenPassingAnIndexReturnsDoubleInThatPosition) {
+    double actual_value= test_vector[0];
+
+    ASSERT_EQ(actual_value, 1.0);
+}
+
+TEST_F(Vector3Test, BracketOperatorWhenPassingAnIndexReturnsDoubleReferenceInThatPosition) {
+    double* expected_value= std::addressof(test_vector.e[0]);
+
+    double* actual_value= std::addressof(test_vector[0]);
+
+    ASSERT_EQ(actual_value, expected_value);
+}
```

Comment:
Parameterized tests are an excellent idea. I added the parameterized test in the latest commit, you'll probably notice that they're at the very end of the file, I did this because it's the order in which they are executed.

I think in this case it's ok to keep them in the same file, but we are creating a new test fixture so if we do end up splitting it we might need to create a directory for all Vector3 tests, we could discuss this in the next meeting.

![TestOrder](https://user-images.githubusercontent.com/17553732/117096229-49475300-ad1d-11eb-9715-42e2faae6904.PNG)